File: frontend/web/src/utils/clockwise-sort.ts
export type Point = { x: number; y: number };

function getAngle(center: Point, point: Point) {
  const x = point.x - center.x;
  const y = point.y - center.y;

  const angle = Math.atan2(y, x);

  if (angle <= 0) {
    return angle + Math.PI * 2;
  }

  return angle;
}

function getDistance(a: Point, b: Point) {
  return Math.sqrt((a.x - b.x) ** 2 + (a.y - b.y) ** 2);
}

function comparePoints(center: Point, a: Point, b: Point) {
  const angleA = getAngle(center, a);
  const angleB = getAngle(center, b);

  if (angleA < angleB) {
    return true;
  }

  const distanceA = getDistance(center, a);
  const distanceB = getDistance(center, b);

  if (angleA === angleB && distanceA < distanceB) {
    return true;
  }

  return false;
}

export function sortPoints(points: Point[], pointCenter: Point) {
  for (const point of points) {
    point.x -= pointCenter.x;
    point.y -= pointCenter.y;
  }

  points = points.sort((a, b) => {
    if (comparePoints({ x: 0, y: 0 }, a, b)) {
      return -1;
    }

    return 1;
  });

  for (const point of points) {
    point.x += pointCenter.x;
    point.y += pointCenter.y;
  }

  return points;
}

File: backend/models/world_map/clockwise_sort.go
package world_map

import (
	"math"
	"sort"
)

func getAngle(center Point, point Point) float64 {
	x := point.X - center.X
	y := point.Y - center.Y

	angle := math.Atan2(float64(y), float64(x))

	if angle <= 0 {
		return angle + math.Pi*2
	}

	return angle
}

func getDistance(a Point, b Point) float64 {
	return math.Sqrt(math.Pow(float64(a.X-b.X), 2) + math.Pow(float64(a.Y-b.Y), 2))
}

func comparePoints(center Point, a Point, b Point) bool {
	angleA := getAngle(center, a)
	angleB := getAngle(center, b)

	if angleA < angleB {
		return true
	}

	distanceA := getDistance(center, a)
	distanceB := getDistance(center, b)

	if angleA == angleB && distanceA < distanceB {
		return true
	}

	return false
}

func Sort[T Pointable](points []T) (center Point, sorted []T) {
	center = Point{X: 0, Y: 0}

	for _, point := range points {
		center.X += point.Point().X
		center.Y += point.Point().Y
	}

	center.X /= len(points)
	center.Y /= len(points)

	sort.SliceStable(points, func(i, j int) bool {
		return comparePoints(center, *points[i].Point(), *points[j].Point())
	})

	return center, points
}
